File: midolmanj-mgmt/src/main/java/com/midokura/midolman/mgmt/host/rest_api/HostInterfacePortResource.java
/*
 * Copyright 2012 Midokura PTE LTD.
 */
package com.midokura.midolman.mgmt.host.rest_api;

import com.google.inject.Inject;
import com.google.inject.assistedinject.Assisted;
import com.google.inject.servlet.RequestScoped;
import com.midokura.midolman.mgmt.ResourceUriBuilder;
import com.midokura.midolman.mgmt.VendorMediaType;
import com.midokura.midolman.mgmt.auth.AuthRole;
import com.midokura.midolman.mgmt.host.HostInterfacePort;
import com.midokura.midolman.mgmt.rest_api.BadRequestHttpException;
import com.midokura.midolman.mgmt.rest_api.NotFoundHttpException;
import com.midokura.midolman.state.StateAccessException;
import com.midokura.midonet.cluster.DataClient;
import com.midokura.midonet.cluster.data.host.VirtualPortMapping;

import javax.annotation.security.RolesAllowed;
import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.UUID;

/**
 * REST API handler for host interface port mapping.
 */
@RequestScoped
public class HostInterfacePortResource {

    private final UUID hostId;
    private final UriInfo uriInfo;
    private final Validator validator;
    private final DataClient dataClient;

    @Inject
    public HostInterfacePortResource(UriInfo uriInfo,
                                     Validator validator,
                                     DataClient dataClient,
                                     @Assisted UUID hostId) {
        this.uriInfo = uriInfo;
        this.validator = validator;
        this.dataClient = dataClient;
        this.hostId = hostId;
    }

    @POST
    @RolesAllowed({AuthRole.ADMIN})
    @Consumes({ VendorMediaType.APPLICATION_HOST_INTERFACE_PORT_JSON,
                   MediaType.APPLICATION_JSON })
    public Response create(HostInterfacePort map)
            throws StateAccessException {

        map.setHostId(hostId);

        Set<ConstraintViolation<HostInterfacePort>> violations =
                validator.validate(map, HostInterfacePort
                        .HostInterfacePortCreateGroup.class);
        if (!violations.isEmpty()) {
            throw new BadRequestHttpException(violations);
        }

        dataClient.hostsAddVrnPortMapping(hostId, map.getPortId(),
                map.getInterfaceName());

        return Response.created(
                ResourceUriBuilder.getHostInterfacePort(uriInfo.getBaseUri(),
                        hostId, map.getPortId()))
                .build();
    }

    @DELETE
    @RolesAllowed({AuthRole.ADMIN})
    @Path("{portId}")
    public void delete(@PathParam("portId") UUID portId)
            throws StateAccessException {

        if (!dataClient.hostsVirtualPortMappingExists(hostId, portId)) {
            return;
        }

        dataClient.hostsDelVrnPortMapping(hostId, portId);
    }

    @GET
    @RolesAllowed({AuthRole.ADMIN})
    @Produces({VendorMediaType
            .APPLICATION_HOST_INTERFACE_PORT_COLLECTION_JSON})
    public List<HostInterfacePort> list() throws StateAccessException {

        List<VirtualPortMapping> mapConfigs =
                dataClient.hostsGetVirtualPortMappingsByHost(hostId);
        List<HostInterfacePort> maps = new ArrayList<HostInterfacePort>();
        for (VirtualPortMapping mapConfig : mapConfigs) {
            HostInterfacePort map = new HostInterfacePort(
                    hostId, mapConfig);
            map.setBaseUri(uriInfo.getBaseUri());
            maps.add(map);
        }

        return maps;
    }

    @GET
    @RolesAllowed({AuthRole.ADMIN})
    @Path("{portId}")
    public HostInterfacePort get(@PathParam("portId") UUID portId)
        throws StateAccessException {

        VirtualPortMapping data = dataClient.hostsGetVirtualPortMapping(
                hostId, portId);
        if (null == data) {
            throw new NotFoundHttpException("Mapping does not exist");
        }

        HostInterfacePort map = new HostInterfacePort(hostId, data);
        map.setBaseUri(uriInfo.getBaseUri());
        return map;
    }
}

File: midonet-cluster/src/main/scala/org/midonet/cluster/services/rest_api/resources/HostInterfacePortResource.scala
/*
 * Copyright 2015 Midokura SARL
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.midonet.cluster.services.rest_api.resources

import java.util.{UUID, List => JList}

import javax.ws.rs._
import javax.ws.rs.core.MediaType.APPLICATION_JSON
import javax.ws.rs.core.Response

import scala.collection.JavaConverters._

import com.google.inject.Inject
import com.google.inject.servlet.RequestScoped

import org.midonet.cluster.rest_api.models.{Host, HostInterfacePort, Port}
import org.midonet.cluster.rest_api.validation.MessageProperty.{HOST_INTERFACE_IS_USED, HOST_IS_NOT_IN_ANY_TUNNEL_ZONE, PORT_ALREADY_BOUND, getMessage}
import org.midonet.cluster.rest_api.{ConflictHttpException, NotFoundHttpException}
import org.midonet.cluster.services.rest_api.MidonetMediaTypes._
import org.midonet.cluster.services.rest_api.resources.MidonetResource.ResourceContext

@RequestScoped
class HostInterfacePortResource @Inject()(hostId: UUID,
                                          resContext: ResourceContext)
    extends MidonetResource[HostInterfacePort](resContext) {

    @GET
    @Path("{id}")
    override def get(@PathParam("id") id: String,
                     @HeaderParam("Accept") accept: String)
    : HostInterfacePort = {
        val portId = UUID.fromString(id)
        getResource(classOf[Host], hostId).portIds.asScala
            .find(_ == portId)
            .map(getResource(classOf[HostInterfacePort], _))
            .getOrElse(throw new NotFoundHttpException("Host interface port not found"))
    }

    @GET
    @Produces(Array(APPLICATION_HOST_INTERFACE_PORT_COLLECTION_JSON))
    override def list(@HeaderParam("Accept") accept: String)
    : JList[HostInterfacePort] = {
        val host = getResource(classOf[Host], hostId)
        listResources(classOf[HostInterfacePort], host.portIds.asScala).asJava
    }

    @POST
    @Consumes(Array(APPLICATION_HOST_INTERFACE_PORT_JSON,
                    APPLICATION_JSON))
    override def create(binding: HostInterfacePort,
                        @HeaderParam("Content-Type") contentType: String)
    : Response = tryTx { tx =>
        val host = tx.get(classOf[Host], hostId)
        if (host.tunnelZoneIds.isEmpty) {
            throw new ConflictHttpException(
                getMessage(HOST_IS_NOT_IN_ANY_TUNNEL_ZONE, hostId))
        }

        val oldPort = tx.get(classOf[Port], binding.portId)
        if (oldPort.interfaceName ne null) {
            throw new ConflictHttpException(
                getMessage(PORT_ALREADY_BOUND, binding.portId))
        }

        tx.list(classOf[Port], host.portIds.asScala)
            // of those ports in the host, see if any has the same ifc
            .find(_.interfaceName == binding.interfaceName) match {
                case Some(conflictingPort) =>
                    throw new ConflictHttpException(
                        getMessage(HOST_INTERFACE_IS_USED,
                                   binding.interfaceName))
                case _ =>
            }

        val port = tx.get(classOf[Port], binding.portId)
        binding.setBaseUri(resContext.uriInfo.getBaseUri)
        binding.create(hostId)
        port.hostId = hostId
        port.interfaceName = binding.interfaceName
        tx.update(port)
        Response.created(binding.getUri).build()
    }

    @DELETE
    @Path("{id}")
    override def delete(@PathParam("id") id: String): Response = tryTx { tx =>
        val port = tx.get(classOf[Port], id)
        port.previousHostId = port.hostId
        port.hostId = null
        port.interfaceName = null
        tx.update(port)
        MidonetResource.OkNoContentResponse
    }

}
